types/multer-s3/index.d.ts

```diff
@@ -3,54 +3,60 @@
-    cacheControl?: ((req: Express.Request, file: Express.Multer.File, callback: (error: any, cacheControl?: string) => void) => void) | string;
-    serverSideEncryption?: ((req: Express.Request, file: Express.Multer.File, callback: (error: any, serverSideEncryption?: string) => void) => void) | string;
-}
-
-declare global {
-    namespace Express {
-        namespace MulterS3 {
-            interface File extends Multer.File {
-                bucket: string;
-                key: string;
-                acl: string;
-                contentType: string;
-                contentDisposition: null;
-                storageClass: string;
-                serverSideEncryption: null;
-                metadata: any;
-                location: string;
-                etag: string;
-            }
-        }
+declare namespace MulterS3 {
+    interface File {
+        /** Field name specified in the form */
+        fieldname: string;
+        /** Name of the file on the user's computer */
+        originalname: string;
+        /** Encoding type of the file */
+        encoding: string;
+        /** Mime type of the file */
+        mimetype: string;
+        /** Size of the file in bytes */
+        size: number;
+        /** The folder to which the file has been saved (DiskStorage) */
+        destination: string;
+        /** The name of the file within the destination (DiskStorage) */
+        filename: string;
+        /** Location of the uploaded file (DiskStorage) */
+        path: string;
+        /** A Buffer of the entire file (MemoryStorage) */
+        buffer: Buffer;
     }
-}
 
-interface S3Storage {
-    (options?: Options): StorageEngine;
+    interface Options<R> {
+        s3: AWS.S3;
+        bucket: ((req: R, file: File, callback: (error: any, bucket?: string) => void) => void) | string;
+        key?(req: R, file: File, callback: (error: any, key?: string) => void): void;
+        acl?: ((req: R, file: File, callback: (error: any, acl?: string) => void) => void) | string;
+        contentType?(req: R, file: File, callback: (error: any, mime?: string, stream?: NodeJS.ReadableStream) => void): void;
+        metadata?(req: R, file: File, callback: (error: any, metadata?: any) => void): void;
+        cacheControl?: ((req: R, file: File, callback: (error: any, cacheControl?: string) => void) => void) | string;
+        serverSideEncryption?: ((req: R, file: File, callback: (error: any, serverSideEncryption?: string) => void) => void) | string;
+    }
 
-    AUTO_CONTENT_TYPE(
-        req: Express.Request,
-        file: Express.Multer.File,
-        callback: (error: any, mime?: string, stream?: NodeJS.ReadableStream) => void): void;
-    DEFAULT_CONTENT_TYPE(
-        req: Express.Request,
-        file: Express.Multer.File,
-        callback: (error: any, mime?: string) => void): void;
+    interface S3Storage {
+        // tslint:disable-next-line:no-unnecessary-generics
+        <E, R>(options?: Options<R>): E;
```

Comment:
Why is the return type not still `StorageEngine`? I don't think the caller can change the fact that a `S3Storage` instance is returned.